rust/src/storage/s3/dynamodb.rs

```diff
@@ -0,0 +1,396 @@
+        Self { client, opts }
+    }
+
+    pub async fn acquire_lock(&self) -> Result<LockItem, DynamoError> {
+        let mut state = AcquireLockState {
+            client: self,
+            active_lock: None,
+            started: Instant::now(),
+            timeout_in: Duration::from_secs(1),
+        };
+
+        loop {
+            match state.try_acquire_lock().await {
+                Ok(lock) => return Ok(lock),
+                Err(DynamoError::ConditionalCheckFailed) => {
+                    println!("got conditiona check, again");
+                    if state.has_timed_out() {
+                        return Err(DynamoError::TimedOut(state.started.elapsed().as_secs()));
+                    }
+                    tokio::time::sleep(Duration::from_secs(1)).await;
+                }
+                Err(e) => return Err(e),
+            }
+        }
+    }
+
+    pub async fn get_lock(&self) -> Result<Option<LockItem>, DynamoError> {
+        let output = self
+            .client
+            .get_item(GetItemInput {
+                consistent_read: Some(true),
+                table_name: self.opts.table_name.clone(),
+                key: hashmap! {
+                    PARTITION_KEY_NAME.to_string() => attr(PARTITION_KEY_VALUE)
+                },
+                ..Default::default()
+            })
+            .await?;
+
+        if let Some(item) = output.item {
+            let get_value = |key| -> Result<String, DynamoError> {
+                Ok(item
+                    .get(key)
+                    .and_then(|r| r.s.as_ref())
+                    .ok_or(DynamoError::InvalidItemSchema)?
+                    .clone())
+            };
+
+            let lease_duration = get_value(LEASE_DURATION)?
+                .parse::<u64>()
+                .map_err(|_| DynamoError::InvalidItemSchema)?;
+
+            return Ok(Some(LockItem {
+                owner_name: get_value(OWNER_NAME)?,
+                record_version_number: get_value(RECORD_VERSION_NUMBER)?,
+                lease_duration,
+                is_released: item.contains_key(IS_RELEASED),
+                data: get_value(DATA).ok(),
+                lookup_time: now_millis(),
+            }));
+        }
+
+        Ok(None)
+    }
+
+    pub async fn release_lock(&self, lock: &LockItem) -> Result<(), DynamoError> {
+        let mut names = hashmap! {
+            vars::PK_PATH.to_string() => PARTITION_KEY_NAME.to_string(),
+            vars::RVN_PATH.to_string() => RECORD_VERSION_NUMBER.to_string(),
+            vars::OWNER_NAME_PATH.to_string() => OWNER_NAME.to_string(),
+            vars::IS_RELEASED_PATH.to_string() => IS_RELEASED.to_string(),
+        };
+        let mut values = hashmap! {
+            vars::IS_RELEASED_VALUE.to_string() => attr("1"),
+            vars::RVN_VALUE.to_string() => attr(&lock.record_version_number),
+            vars::OWNER_NAME_VALUE.to_string() => attr(&lock.owner_name),
+        };
```

Comment:
@houqp there's several other places where hashmap is created with k/v `k.to_string() -> y.to_string()` and `k.to_string -> attr(v)`

For the sake of clearer code, does it make sense to just use `k -> v` and then apply some mapping function `convert(hashmap!{k -> v})`;
does performance penalty (of creating second hash map) worth it?

Reply:
We can get what you want with zero overhead by creating our own helper macros, for example:

```rust
macro_rules! string_hashmap {
    ($($key:expr => $value:expr,)+) => { hashmap!($($key.to_string() => $value),+) };
}

fn main() {
    let mut m = string_hashmap!{
        "a" => 1,
        "b" => 2,
    };
    m.insert("c".to_string(), 3);

    dbg!(m);
}
```

output:

```
   Compiling playground v0.0.1 (/playground)
    Finished dev [unoptimized + debuginfo] target(s) in 2.50s
     Running `target/debug/playground`
[src/main.rs:31] m = {
    "b": 2,
    "c": 3,
    "a": 1,
}
```

Reply:
neat